### PR TITLE
Use the ParDo Application to Cache DoFns

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

A DoFn application is the scope of reuse.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoSingleEvaluatorFactory.java [hunk 1]
```diff
@@ -55,27 +69,38 @@ public ParDoSingleEvaluatorFactory() {
...
+              Collections.<TupleTag<?>>emptyList(),
+              ImmutableMap.<TupleTag<?>, PCollection<?>>of(mainOutputTag, application.getOutput()));
       return ThreadLocalInvalidatingTransformEvaluator.wrapping(parDoEvaluator, fnLocal);
     } catch (Exception e) {
       fnLocal.remove();
       throw e;
     }
   }
+
+  private static class BoundToFn
```
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoSingleEvaluatorFactory.java]: unused?

**Comment by kennknowles** [comment]
Conceptually LGTM. An important fix.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoSingleEvaluatorFactory.java (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoSingleEvaluatorFactory.java]: Yes. Rm'd.

**Pull request closed without merging**